pkg/apis/kubeone/v1beta1/defaults.go

```diff
@@ -118,13 +120,25 @@ func SetDefaults_Versions(obj *KubeOneCluster) {
 }
 
 func SetDefaults_ContainerRuntime(obj *KubeOneCluster) {
+	defaultContainerRuntime := ContainerRuntimeConfig{
+		Docker: &ContainerRuntimeDocker{},
+	}
+
+	k121OrMore, _ := semver.NewConstraint(">=1.21")
+	actualVer, err := semver.NewVersion(obj.Versions.Kubernetes)
+
+	if err == nil {
+		if k121OrMore.Check(actualVer) {
+			defaultContainerRuntime.Containerd = &ContainerRuntimeContainerd{}
+			defaultContainerRuntime.Docker = nil
+		}
+	}
```

Comment:
It seems much better alternative will be:
* always default container-runtime to docker, preserving behaviour
* always require containerd to be explicitly configured as container-runtime
* fail validation for kubernetes 1.22+ if docker is indicated as container-runtime

P.S.
flatcar is working on containerd support at the moment, but it's not yet there.

Reply:
The aforesaid is to avoid any "magic" jumps in logic and behaviour and keep being predictable.

Reply:
The idea behind switching to containerd for the new clusters as of 1.21 is to encourage users to try and test it before we (and Kubernetes) make it mandatory as of 1.22. Of course, users could do it manually, but if we will already do it, the idea was to do it earlier (in this case 1.21).

Also, the idea was that users don't have to change their manifest (keeping backward compatibility), so if a cluster is already provisioned with Docker, we would use Docker while it's supported.

Another case that came to my mind is that if some user would stamp a custom image that has Docker preinstalled, and we default to containerd, then we might/would fail to provision the cluster. The same applies in the other direction, i.e. someone has containerd preinstalled and we try to use Docker.

I think the best scenario would be that we determine the container runtime dynamically like we determine the hostname and the operating systems. Determining container runtime would happen right after those two tasks and it would work in the following way:

* If container runtime is configured in the manifest, skip determining the CR
* Check is there Docker or containerd already installed and default to the installed CR
* If nothing is installed:
  * For Kubernetes 1.21 and newer: use containerd
  * For Kubernetes 1.20 and older: use Docker
* At the end, validate is the configuration supported (we already have that, so nothing to do from that side)

I think that this should be relatively easy to implement and would solve all our problems, while we would still be able to follow the original plan.